presale/models/presale_order.py

```diff
@@ -0,0 +1,89 @@
+from odoo import api, fields, models
+
+
+class PresaleOrder(models.Model):
+    _name = 'presale.order'
+    _description = 'presales orders'
+
+    name = fields.Char(string='Preorder Reference',
+                       required=True, copy=False, readonly=True,
+                       default=lambda self: 'New')
+
+    active = fields.Boolean('Active', default=True)
+
+    state = fields.Selection(
+        selection=[
+            ('draft', 'Draft'),
+            ('confirmed', 'Confirmed')
+        ],
+        string='Status',
+        readonly=True, copy=False,
+        default='draft'
+    )
+
+    partner_id = fields.Many2one(
+        comodel_name='res.partner',
+        string="Customer",
+        required=True, readonly=False, change_default=True)
+
+    order_lines = fields.One2many(
+        'presale.order.line', 'order_id', string='Order Lines')
+
+    sale_order_id = fields.Many2one(
+        'sale.order', string='Sale Order', copy=False)
+
+    def validate_action(self):
+        return True
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get('name', "New") == "New":
+                vals['name'] = self.env['ir.sequence'].next_by_code(
+                    'presale.order') or "New"
+
+        return super().create(vals_list)
+
+    def validate_action(self):
+        SaleOrder = self.env['sale.order']
+        MailTemplate = self.env['mail.template']
+
+        # Create sale order
+        sale_order_vals = {
+            'partner_id': self.partner_id.id,
+            'presale_order_id': self.id
+            # Other sale order values
+        }
+        sale_order = SaleOrder.create(sale_order_vals)
+
+        # Define SaleOrderLine model
+        SaleOrderLine = self.env['sale.order.line']
+        for line in self.order_lines:
+            sale_order_line_vals = {
+                'order_id': sale_order.id,
+                'product_id': line.product_id.id,
+                'product_uom_qty': line.quantity,
+                'price_unit': line.price,
+                # Other sale order line values
+            }
+            sale_order_line = SaleOrderLine.create(sale_order_line_vals)
```

Comment:
You can create order lines directly during the creation of the order, this is more efficient as it will do everything in a single SQL transaction and has the benefit of rolling back the whole order if the creation of one of the lines fails
```suggestion
        # Create sale order
        sale_order_line_vals = [
            {
                'product_id': line.product_id.id,
                'product_uom_qty': line.quantity,
                'price_unit': line.price,
                # Other sale order line values
            }
            for line in self.order_lines
        ]
        
        sale_order_vals = {
            'partner_id': self.partner_id.id,
            'presale_order_id': self.id,
            'order_line': Command.create(sale_order_line_vals),
            # Other sale order values
        }
        
        sale_order = SaleOrder.create(sale_order_vals)
```

Reply:
if we are going to create the sale_order_line before the sale_order how is the order_line will be linked to the sale_order without adding the sale_order id to it?